### PR TITLE
EDA part 2: core principles.

### PR DESCRIPTION
**Pull request by rtxm**

### CODE REVIEW

File: _posts/2020-03-04-eda-architecture2.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+---
+layout: post
+title:  "An event driven architecture — part 2"
+excerpt: "Core design principles."
+date:   2020-03-04 12:53:00 +0100
+categories: architecture
+tags:  services event bus architecture
+---
+
+In the previous post in this series, we explained why we ditched our old architecture based on synchronous REST services for a completely asynchronous event driven architecture.
```
Comment (shostelet) [inline _posts/2020-03-04-eda-architecture2.md]: Will it be easy to add a link to ease navigation from article to article?

I would also suggest, at the end of each article, listing every articles and update the text to a link once a given article has been published.

File: _posts/2020-03-04-eda-architecture2.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+In the same way we have two different kinds of services, we have two different kinds of messages: *Commands* and *Events*.
+
+Events are business messages published on the bus by business services and telling the world what happened.
+
+A business service owns the type of Events it emits. It knows nothing about the services who will process them. It subscribes to the types of Events it needs but knows nothing about their origins.
+
+An Event type defines the meaning of the events of that type and their data schema. They must be documented.
+
+The type of an actual event message is given by its name (a.k.a. *routing key* because it is used by the subscription routing). The event type name must be in the form `SubjectPastParticiple`. For example, `ConversationStarted`, `CustomerCreated`, `ShoppingCartValidated`… If you're not able to immediately give a name to your event type, it means it is not well defined, or that it is not an event. Maybe, you need to refine your service or split it, as you may not have analyzed your value chain deeply enough?
```
Comment (shostelet) [inline _posts/2020-03-04-eda-architecture2.md]: I don't know the rule regarding ellipsis following a comma or not, but you've written `(computations, transformations, …)` above, but ``ConversationStarted`, `CustomerCreated`, `ShoppingCartValidated`…` here.
We might consider formatting with the same rule.

File: _posts/2020-03-04-eda-architecture2.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+As commands are addressed to a particular logical service, the *routing key* of a command is in the form `logical_service_name.commandname`. For example: `voxo.TranscribeAudio`.
+
+The command contains the return "address" to which the result is to be sent and a reference set by the sender that is returned as-is, along with the command outcome. That reference is called the *correlation identifier* and it is very important for the sender: as all communications are asynchronous, the service requesting the command needs a way to reconcile the received result with the initial request it made.
+
+A Result contains the result of the process, that can be the successful outcome, or an error, and the correlation identifier.
+
+Error results are expected and documented: they are “normal” errors, not bug reports. Bug exceptions *must not return an error result*. In case of unexpected error, the service will replace the input command on the queue once, and if a second try raises an unexpected error again, the message is refused and goes into the dead letter queue for investigation. The exceptions are always logged.
```
Comment (shostelet) [inline _posts/2020-03-04-eda-architecture2.md]: `the service will replace the input command on the queue once` did you meant `will retry the input command`?
If you really meant `replace`, I'm not sure I understand what you mean.

File: _posts/2020-03-04-eda-architecture2.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+We can also have logging messages, on their own exchange (see below), to easily collect application logs.
+
+All messages contain a conversation identifier which has the following properties:
```
Comment (shostelet) [inline _posts/2020-03-04-eda-architecture2.md]: Might be useful to specify that the concept of "conversation" is specific to our business?

File: _posts/2020-03-04-eda-architecture2.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+In the previous post in this series, we explained why we ditched our old architecture based on synchronous REST services for a completely asynchronous event driven architecture.
+
+Today, we address the core design principles that were crucial in the success of this enterprise.
+
+## Business Services and Data Processing Services
+
+We make a distinction between *Business Services* and *Data Processing Services* (a.k.a. utility services) to cleanly separate business logic from data processing complexity.
+
+DP Services are expected to be  **pure, stateless, services** that provide some kind of algorithmic data processing (computations, transformations, …). Moreover, they are also context free: they should not depend on business rules, assumptions or external data sources. All they need to do their processing must be in the message they receive. They should not have to query a tier to get more data. DP services are kind of *universal* libraries and can even be provided by tiers.
```
Comment (shostelet) [inline _posts/2020-03-04-eda-architecture2.md]: Could you illustrate Business Services and Data Processing Services with examples from our architecture, as well as maybe examples from a easily understandable point of reference? (such as an e-commerce example, that would echo the event `ShoppingCartValidated` below)

File: _posts/2020-03-04-eda-architecture2.md (hunk 4 above)
Comment (rtxm) [reply _posts/2020-03-04-eda-architecture2.md]: That's not *that*  conversation. Nothing to do with speech. I'll explain more.

**Comment by vjousse** [comment]
Moved to: https://github.com/allo-media/allo-media.github.io/pull/80

**Pull request closed without merging**